core/src/main/java/org/polypheny/db/catalog/snapshot/impl/SnapshotImpl.java

```diff
@@ -0,0 +1,333 @@
+import org.polypheny.db.catalog.entity.LogicalUser;
+import org.polypheny.db.catalog.entity.logical.LogicalCollection;
+import org.polypheny.db.catalog.entity.logical.LogicalEntity;
+import org.polypheny.db.catalog.entity.logical.LogicalGraph;
+import org.polypheny.db.catalog.entity.logical.LogicalNamespace;
+import org.polypheny.db.catalog.entity.logical.LogicalTable;
+import org.polypheny.db.catalog.logistic.Pattern;
+import org.polypheny.db.catalog.snapshot.AllocSnapshot;
+import org.polypheny.db.catalog.snapshot.LogicalDocSnapshot;
+import org.polypheny.db.catalog.snapshot.LogicalGraphSnapshot;
+import org.polypheny.db.catalog.snapshot.LogicalRelSnapshot;
+import org.polypheny.db.catalog.snapshot.Snapshot;
+import org.polypheny.db.iface.QueryInterfaceManager.QueryInterfaceTemplate;
+
+@Value
+@Accessors(fluent = true)
+public class SnapshotImpl implements Snapshot {
+
+
+    @NotNull
+    LogicalRelSnapshot rel;
+    @NotNull
+    LogicalDocSnapshot doc;
+    @NotNull
+    LogicalGraphSnapshot graph;
+    @NotNull
+    AllocSnapshot alloc;
+
+    @Getter
+    long id;
+
+    ImmutableMap<Long, LogicalUser> users;
+
+    ImmutableMap<String, LogicalUser> userNames;
+    ImmutableMap<Long, LogicalQueryInterface> interfaces;
+    ImmutableMap<String, QueryInterfaceTemplate> interfaceTemplates;
+
+    ImmutableMap<String, LogicalQueryInterface> interfaceNames;
+    ImmutableMap<Long, LogicalAdapter> adapters;
+    ImmutableMap<Long, AdapterTemplate> adapterTemplates;
+
+    ImmutableMap<String, LogicalAdapter> adapterNames;
+
+    ImmutableMap<Long, LogicalNamespace> namespaces;
+
+    ImmutableMap<String, LogicalNamespace> namespaceNames;
+
+
+    public SnapshotImpl( long id, Catalog catalog, Map<Long, LogicalNamespace> namespaces, LogicalRelSnapshot rel, LogicalDocSnapshot doc, LogicalGraphSnapshot graph, AllocSnapshot alloc ) {
+        this.id = id;
+        this.rel = rel;
+        this.doc = doc;
+        this.graph = graph;
+
+        this.namespaces = ImmutableMap.copyOf( namespaces );
+
+        this.namespaceNames = ImmutableMap.copyOf( namespaces.values().stream().collect( Collectors.toMap( n -> n.caseSensitive ? n.name : n.name.toLowerCase(), n -> n ) ) );
+
+        this.alloc = alloc;
+
+        this.users = ImmutableMap.copyOf( catalog.getUsers() );
+        this.userNames = ImmutableMap.copyOf( users.values().stream().collect( Collectors.toMap( u -> u.name, u -> u ) ) );
+        this.interfaces = ImmutableMap.copyOf( catalog.getInterfaces() );
+        this.interfaceNames = ImmutableMap.copyOf( interfaces.values().stream().collect( Collectors.toMap( i -> i.name, i -> i ) ) );
+        this.interfaceTemplates = ImmutableMap.copyOf( catalog.getInterfaceTemplates() );
+        this.adapters = ImmutableMap.copyOf( catalog.getAdapters() );
+        this.adapterNames = ImmutableMap.copyOf( adapters.values().stream().collect( Collectors.toMap( a -> a.uniqueName, a -> a ) ) );
```

Comment:
The same argument regarding case-sensitivity can be made for users, adapters and interfaces